call_chaos/call_chaos/src/lib.rs

```diff
@@ -0,0 +1,394 @@
+    OnewayError, RejectCode, Response,
+};
+use lazy_static::lazy_static;
+use rand::{Rng, SeedableRng};
+use rand_chacha::ChaCha8Rng;
+use std::future::IntoFuture;
+use std::mem;
+use std::pin::Pin;
+use std::sync::Mutex;
+use std::task::Poll;
+
+/// A trait that defines a policy for allowing or rejecting calls.
+pub trait Policy: Send + Sync {
+    /// Whether to allow a call.
+    /// If the call is allowed, the ic_cdk call will be executed. This still doesn't mean that the
+    /// call will succeed, as it might actually fail for an arbitrary reason (e.g., not having enough cycles,
+    /// system being under load, etc).
+    /// If the call is not allowed, the ic_cdk call will not be executed. The policy can, however,
+    /// execute the call under the hood and still return an error. For example, for bounded-wait
+    /// calls, you may return `RejectCode::SysUnknown` and still issue the call, simulating a
+    /// timeout in production. Returning a `CallFailed::CallRejected` will cause the chaos library
+    /// to actually delay producing the rejection, by calling a no-op on the management canister.
+    /// This is done in order to simulate the time it would take for the call to be rejected in
+    /// production.
+    /// Note that this takes a mutable reference to the policy, so it can be used to maintain state
+    /// if needed (e.g., drop the first `N` calls, and then allow all calls to go through)
+    fn allow(&mut self, call: &Call) -> Result<(), CallFailed>;
+
+    /// Whether to allow a one-way call.
+    /// If the call is allowed, the ic_cdk call will be executed. This still doesn't mean that the
+    /// call will succeed, as it might fail for an arbitrary reason (e.g., not having enough cycles,
+    /// system being under load, etc).
+    /// If the call is not allowed, the ic_cdk call will not be executed. To allow simulating the
+    /// call failing silently, the error is returned as an `Option<OnewayError>`. An error of
+    /// `None` means that the call shouldn't be executed but no error should be returned either. An
+    /// error of `Some(OnewayError)` means that the call shouldn't be executed and the error should
+    /// be returned to the caller.
+    fn allow_oneway(&mut self, call: &Call) -> Result<(), Option<OnewayError>>;
+}
+
+/// A simple policy that allows all calls.
+#[derive(Default)]
+pub struct AllowAll {}
+
+impl Policy for AllowAll {
+    fn allow(&mut self, _call: &Call) -> Result<(), CallFailed> {
+        Ok(())
+    }
+
+    fn allow_oneway(&mut self, _call: &Call) -> Result<(), Option<OnewayError>> {
+        Ok(())
+    }
+}
+
+/// A simple policy that denies all calls, returning a `SysTransient` reject code.
+#[derive(Default)]
+pub struct DenyAll {}
+
+impl Policy for DenyAll {
+    fn allow(&mut self, _call: &Call) -> Result<(), CallFailed> {
+        Err(CallFailed::CallRejected(CallRejected::with_rejection(
+            RejectCode::SysTransient as u32,
```

Comment:
Wondering whether it makes sense to make the reject code configurable (e.g. as a field of `DenyAll`. Maybe even a list of reject codes to pick from, randomly or round-robin.

Reply:
We could, but depends on what you want to test. In general, a non-`SysTransient` is either a potentially "unclean" reject, or a `DestinationInvalid`, which is not retryable, so everything that does retries would handle these reject codes differently. For the moment I'd leave it as-is, noting that the user can implement their own `Policy` anyways if they want more functionality. If anyone actually starts using this library then we can revisit based on the feedback :)

It's a similar choice for all the other deny policies.